spellbook.lic

```diff
@@ -0,0 +1,196 @@
+  end
+
+  # TODO: would be nice if this was a wrapper that always did something and put the spellbook away safely.
+  # TODO: this would prevent me from being in an bad state if the script doesn't exit
+  # TODO: and if the script does exit unexpectedly, then before_dying would call the any unhandled exit.
+  def safe_get_spellbook
+    safe_empty_hands
+    open_and_turn_spellbook_case_to_spellbook
+    if check_indoors
+      bput("get my #{@spellbook_name} from my #{@spellbook_case}", "You get")
+    else
+      echo("*** WARNING: You must be indoors! This is very unsafe! Please debug this before using the script again... exiting.")
+      safe_exit
+    end
+  end
+
+  def store_spellbook
+    open_and_turn_spellbook_case_to_spellbook
+    bput("close my #{@spellbook_name}", "You close your", "That is already closed!")
+    tucked_spellbook = bput("put my #{@spellbook_name} in my #{@spellbook_case}", "You tuck")
+    check_spellbook_safe
+    # TODO: verify spellbook case does not need closed
+  end
+
+  def check_spellbook_safe
+    not_found = "I could not find what you were referring to."
+    found = "You tap a"
+    open_and_turn_spellbook_case_to_spellbook
+    result = bput("tap my #{@spellbook_name} in my #{@spellbook_case}", found, not_found)
+    if found == result
+      return true
+    end
+    if not_found == result
+      echo("*** WARNING: SPELLBOOK #{@spellbook_name} NOT FOUND IN #{@spellbook_case}")
+      return false
+    end
+    # TODO: verify spellbook case does not need closed
+  end
+
+  def safe_exit
+    if check_holding_spellbook
+      store_spellbook
+    end
+    exit
+  end
+
+  def open_spellbook_for_casting
+    already_open = "That is already open!"
+    opening_spellbook = bput("open my #{@spellbook_name}", "You open a", already_open)
+    if opening_spellbook = already_open
+      bput("turn my #{@spellbook_name} to contents", "You flip back", "The spellbook is already opened to the table of contents.")
+    end
+
+    bput("turn my #{@spellbook_name} to chapter #{@spellbook_scrolls_chapter}", "You turn the")
+  end
+
+  def find_spell_page
+    # TODO: consider trying this 3+ times
+    bput("read my #{@spellbook_name}", "spellbook reads")
+    page_line = matchtimeout 5, @spellbook_spell
+    if not page_line
+      echo("*** WARNING: The spell is not in your spellbook. Did it get erased?")
+      safe_exit
+    end
+    # ref: Page 4.......Iron Constitution
+    page_number = page_line[/\d+/]  # get first decimal substring
+    return page_number
+  end
+
+  # TODO NB: It is not necessary to read the spellbook to unleash it.
+  def turn_to_spell_page(page_number)
+    bput("turn my #{@spellbook_name} to page #{page_number}", "You turn to page", "is already open to page")
+  end
+
+  def unleash_spellbook(data)
+    cast_spell(data, @settings)
```

Comment:
Cast unleash _then_ cast the scroll spell?

Reply:
Unfortunately the `cast spellbook` action of unleash doubles as the prep action for the next spell. Because of the way common-arcana.lic is written, it would require some disruptive state tracking and/or rewriting to make a cast action of 1 spell the same action as the prep action for the next spell. Each has a lot of contexts.